tests/scripts/arp_responder.py

```diff
@@ -86,27 +94,44 @@ def __init__(self, ip_sets):
 
     def action(self, interface):
         data = interface.recv()
-        if len(data) >= self.ARP_PKT_LEN:
+        if len(data) > self.ARP_PKT_LEN:
             return
 
-        remote_mac, remote_ip, request_ip = self.extract_arp_info(data)
+        remote_mac, remote_ip, request_ip, op_type, vlan_id = self.extract_arp_info(data)
+
+        # Don't send ARP response if the ARP op code is not request
+        if op_type != self.ARP_OP_REQUEST:
+            return
 
         request_ip_str = socket.inet_ntoa(request_ip)
         if request_ip_str not in self.ip_sets[interface.name()]:
             return
-
-        if 'vlan' in self.ip_sets[interface.name()]:
-            vlan_id = self.ip_sets[interface.name()]['vlan']
-        else:
-            vlan_id = None
-
         arp_reply = self.generate_arp_reply(self.ip_sets[interface.name()][request_ip_str], remote_mac, request_ip, remote_ip, vlan_id)
         interface.send(arp_reply)
 
         return
         
     def extract_arp_info(self, data):
-        return data[6:12], data[28:32], data[38:42] # remote_mac, remote_ip, request_ip
+        # remote_mac, remote_ip, request_ip, op_type
+        rem_ip_start = 28
+        req_ip_start = 38
+        op_type_start = 20
+        eth_offset = 0
+        vlan_id = None
+        ether_type = str(data[12:14]).encode("HEX")
+        if (ether_type == '8100'):
+            vlan = str(data[14:16]).encode("HEX")
+            if (vlan != '0000'):
+                eth_offset = 4
+                vlan_id = data[14:16]
+        rem_ip_start = rem_ip_start + eth_offset
+        req_ip_start = req_ip_start + eth_offset
+        op_type_start = op_type_start + eth_offset
+        rem_ip_end = rem_ip_start + 4
+        req_ip_end = req_ip_start + 4
+        op_type_end = op_type_start + 1
```

Comment:
It would be a lot cleaner to use scapy to parse the packet as an ARP request and check the fields. (I'm guessing all this manual work is here right now b/c `recv` is receiving the raw bytes.)

Reply:
backlog